### PR TITLE
Notes feature

### PR DESCRIPTION
**Pull request by perlmonger42**
Adds documentation of all the new API entry points created to support Notes:
- `POST /<notables>/:id/notes`
- `GET /<notables>/:id/notes`
- `GET /rules/:rule_id/rule_component_notes`
- `GET /notes/:note_id`

for all values of `<notables>` in {`data_elements`, `extensions`, `libraries`, `properties`, `rule_components`, `rules`}.

Fitting this documentation into the current document structure presents some questions. For example, the Create, Fetch, and List operations for Notes could have been nested under the top-level entries for the Notable Types. Using Property as the example, this doesn't look appealing to me (the additions for Notes are indicated in **bold**):

* Property
  * Create
  * **Create Note**
    * **for Data Element**
    * **for Extension**
    * _etc._
  * Delete
  * Fetch
  * **Fetch Note**
  * List
  * **List Notes**
    * **for Data Element**
    * **for Extension**
    * _etc._
  * Update
  * Relationships
    * Callbacks
    * Company
    * DataElements
    * ...
    * Rules

They could conceivably be pushed under Relationships. I don't think they belong there, since Notes have a stronger connection than Relationships typically imply -- Notes are *owned* by their Notable, not merely *related* to them. If we ignore that concern, we get something like this:

* Property
  * Create
  * Delete
  * Fetch
  * List
  * Update
  * Relationships
    * Callbacks
    * Company
    * DataElements
    * **Notes**
      * **Create**
      * **Fetch**
      * **List**
    * ...
    * Rules

Among its other problems, that organization has 6 copies of exactly the same Fetch documentation.

The organization I settled on makes Notes a top-level entity, at the same hierarchical level as DataElements, Rules, etc. This seemed the best solution to me, but I'm not wedded to it, and suggestions are welcomed.

* Audit Events
* Builds
* ...
* Libraries
* **Notes**
  * **Overview**
  * **Fetch**
  * **Create**
    * **DataElement Note**
    * **Extension Note**
    * **Library Note**
    * **Property** Note**
    * **Rule Note**
    * **RuleComponent Note**
  * **List**
    * **DataElement Notes**
    * **Extension Notes**
    * **Library Notes**
    * **Property Notes**
    * **Rule Notes**
    * **RuleComponent Notes**
    * **RuleComponent Notes for Rule**
* Profiles
* Properties
* _etc._

#### Changes
New versions of `endpoints.json`, `error_codes.json`, `filters.json`, and `forms.json` are included with this work. They were generated in a Blacksmith container, on the `notes-feature` branch, using the command `rake api_spec:all`.

### CODE REVIEW

File: api/reference/1.0/notes/index.md
Comment (markhicken) [inline api/reference/1.0/notes/index.md:42]: @thebenrob I can see a potential need for adding notes to an old revision. Perhaps you forgot to document a change when it was made last week and now you want to document it because you are handing off some work to another person. Do we care to support this type of use case?

File: api/reference/1.0/notes/index.md
Comment (markhicken) [inline api/reference/1.0/notes/index.md:51]: @thebenrob I realize this is feature parity with DTM, but I question if historical notes should be editable. As you build out an implementation, you might make changes to Rule A which depends on DataElement A. If you can't revise your notes on DataElement A, you now have to edit it and cut a revision. This seems unnecessary just to add some documentation.

Comment (spencerroan) [reply api/reference/1.0/notes/index.md:51]: Revisions don't have notes, when you ask for a revision's notes you get those from head before created before the revision's creation. This means once a revision is cut, you cannot add notes to it, just to what the next revision will be. You also cannot add notes to a revision, like `This is the revision where super bad bug was introduced`.

Head can have any number of notes, so you don't need to cut a new revision per se.
The way we've approximated notes for revisions is really weird to me. As a user I can't necessarily do the things I would like to do.  If we added notes to revisions, then displaying them becomes problematic because you have 2 time axis. One for revision creation date and one for note creation date. I'd like to see all of the notes for some origin, do I see the notes in time order and show revisions out of order? Do I show notes primarily in the order of the revision, and second by create date of the note?  
If I cut a revision of a rule, add the rule to the library. Then submit the library. The approver wants to say, `this is borked, yo! Change the implementation of x`. That note can only apply to head, which could be very different from the revision that was submitted.

Comment (markhicken) [reply api/reference/1.0/notes/index.md:51]: Yah, I think it might be a nice feature to make notes editable for a revision, but I don't think it's the end of the world not to have it. You can always add a new note to head that says something like `I broke purchase tracking in revision 3 and fixed it in revision 4`. 

Just wanted to bring this up in case we deem it a feature worth having.

File: .gitignore
Comment (remear) [inline .gitignore]: This file shouldn't be ignored.

File: api/reference/1.0/notes/create/data_element.md
Comment (remear) [inline api/reference/1.0/notes/create/data_element.md]: `{% form note.create %}`

File: api/reference/1.0/notes/create/extension.md
Comment (remear) [inline api/reference/1.0/notes/create/extension.md]: `{% form note.create %}`

File: api/reference/1.0/notes/create/library.md
Comment (remear) [inline api/reference/1.0/notes/create/library.md]: `{% form note.create %}`

File: api/reference/1.0/notes/create/property.md
Comment (remear) [inline api/reference/1.0/notes/create/property.md]: `{% form note.create %}`

File: api/reference/1.0/notes/create/rule.md
Comment (remear) [inline api/reference/1.0/notes/create/rule.md]: `{% form note.create %}`

File: api/reference/1.0/notes/create/rule_component.md
Comment (remear) [inline api/reference/1.0/notes/create/rule_component.md]: `{% form note.create %}`

File: api/reference/1.0/notes/index.md
Comment (remear) [inline api/reference/1.0/notes/index.md:13]: Alphabetize these?

File: api/reference/1.0/notes/index.md
Comment (remear) [inline api/reference/1.0/notes/index.md]: @thebenrob Would you please share your thoughts on this section.

File: api/reference/1.0/notes/index.md
Comment (remear) [inline api/reference/1.0/notes/index.md:57]: This section is redundant. The notable resources were already outlined at the start of the document.

File: api/reference/1.0/notes/list/data_element.md
Comment (remear) [inline api/reference/1.0/notes/list/data_element.md]: These list examples should follow a hierarchy of `/1.0/{RESOURCE}/list_notes.md`

**Review by remear — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (perlmonger42) [reply .gitignore]: 👍

File: api/reference/1.0/notes/index.md
Comment (perlmonger42) [reply api/reference/1.0/notes/index.md:13]: I was sorting them from simplest to most complex, but... okay.

File: api/reference/1.0/notes/index.md
Comment (perlmonger42) [reply api/reference/1.0/notes/index.md:57]: Yes. But this section is the *Limitations* section, and I wanted to emphasize here that you can't just drop Notes on _every_ kind of thing, and make it crystal clear to the user that his favorite note target was not just overlooked. Kind of a pre-emptive answer to "Why can't I get the API to let me add Notes to my Environment?"

With that understanding, do you still want me to pull it?

File: api/reference/1.0/notes/list/data_element.md
Comment (perlmonger42) [reply api/reference/1.0/notes/list/data_element.md]: I'll reorganize them.

So, you're okay with all the `create` operations nested under Note, but you want the `list` operations listed under their Notable? I covered a lot of permutations in my PR comment, but that ain't one of them.

So...
* the `create` ops create Notes, so they belong under Notes
* the `list` ops are accessed through a <notable>, so they belong under their Notable

Is that the idea?

File: api/reference/1.0/notes/create/data_element.md
Comment (perlmonger42) [reply api/reference/1.0/notes/create/data_element.md]: Sure 'nuff.

File: api/reference/1.0/notes/list/data_element.md
Comment (remear) [reply api/reference/1.0/notes/list/data_element.md]: That's the idea, for now.

File: api/reference/1.0/notes/index.md
Comment (remear) [reply api/reference/1.0/notes/index.md:57]: I would remove it. Check with @thebenrob.

File: api/reference/1.0/notes/index.md
Comment (thebenrob) [reply api/reference/1.0/notes/index.md]: The current feature set was modeled after DTM in the name of "feature parity".  I think you may be right that some will ask for this, but it raises a number of complications that would have to be dealt with and I'm willing to wait and see what the users think.

File: api/reference/1.0/notes/index.md
Comment (thebenrob) [reply api/reference/1.0/notes/index.md]: I think it's fine.  The last sentence I might change for stylistic reasons and say something like "The contents of `Notes` have no impact on the behavior of property resources, so you are free to use them any way you want."

File: api/reference/1.0/notes/index.md
Comment (thebenrob) [reply api/reference/1.0/notes/index.md:42]: I think not right now.  The current set of functionality was done in the name of "feature parity" with DTM.  I understand the use case and I think you could be right that some will want it, but it would introduce a number of complexities that we didn't want to deal with in the first iteration.  One example, if you try to add a note to head, do we attach it to head and then later attach it to the next revision?  Do we just cut a revision for you and put it there?  If Notes are on revisions, how would I see all notes for any revision?
In any case, I consider this a releasable chunk and we can iterate later based on user feedback.

File: api/reference/1.0/notes/index.md
Comment (thebenrob) [reply api/reference/1.0/notes/index.md:57]: This page is kinda functioning as the Guides section for now and will be the only place in docs (until the UI is done) where we describe the functionality here.  I think it's fine to be explicit about what resources they cannot be attached to.  I would say that we should make sure the list of non-Notable resources does not include any resources that are not already in the documentation (not sure on Profiles for example).

Comment (perlmonger42) [reply api/reference/1.0/notes/index.md:57]: [Profiles](https://developer.adobelaunch.com/api/reference/1.0/profiles/) are already in the documentation. And I've double-checked the rest, and believe they are all documented types.

I actually built that list by reading the [Reference/1.0](https://developer.adobelaunch.com/api/reference/1.0/profiles/#) menu choices out of my local Jekyll instance -- which, surprisingly, caused me to accidentally include 'Adapters' originally. There was an empty directory named 'adapters' on my system -- a holdover from before the rename to 'hosts' -- and that  caused an empty entry in the docs menu, which I blindly copied without thinking. So I carefully reviewed the list in my proposed documentation to make sure they were all legit.

File: api/reference/1.0/notes/index.md
Comment (perlmonger42) [reply api/reference/1.0/notes/index.md]: I've replaced my last sentence with @thebenrob's. And then used that to replace the paragraph _before_ the list of suggestions, since it said virtually the same thing and there was need to repeat it.

File: api/reference/1.0/notes/index.md
Comment (markhicken) [reply api/reference/1.0/notes/index.md:42]: 👍

**Review by remear — DISMISSED** [review]
The page titles might need some refinement later. Let's roll with this for now. Merge this after Notes have been released.

**Review by jeffreywalter — APPROVED** [review]

**Review by jeffreywalter — DISMISSED** [review]

**Review by jbarchibald — DISMISSED** [review]

**Review by jbarchibald — APPROVED** [review]

**Review by thebenrob — APPROVED** [review]

**Pull request merged**
